astronomy/date_body.hpp

```diff
@@ -0,0 +1,781 @@
+  constexpr TimeOfDay const& checked() const;
+
+  int const hour_;
+  int const minute_;
+  int const second_;
+  int const nanosecond_;
+};
+
+class DateTime {
+ public:
+  constexpr Date const& date() const;
+  constexpr TimeOfDay const& time() const;
+
+  // If |time()| is 24:00:00, returns an equivalent DateTime where midnight is
+  // expressed as 00:00:00 on the next day; otherwise, returns *this.
+  constexpr DateTime normalized_end_of_day() const;
+
+ private:
+  constexpr DateTime(Date const date, TimeOfDay const time);
+
+  // Checks that |time| does not represent a leap second unless |date| is the
+  // last day of June, December, March, or September.
+  constexpr DateTime const& checked() const;
+
+  Date const date_;
+  TimeOfDay const time_;
+
+  friend constexpr DateTime operator""_DateTime(char const* string,
+                                                 std::size_t size);
+};
+
+// Arithmetico-calendrical utility functions.
+
+constexpr std::array<int, 12> non_leap_year_month_lengths{
+    {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}};
+
+constexpr bool is_gregorian_leap_year(int const year) {
+  return (year % 4 == 0 && (year % 100 != 0 || year % 400 == 0));
+}
+
+constexpr int gregorian_year_length(int const year) {
+  return is_gregorian_leap_year(year) ? 366 : 365;
+}
+
+constexpr int month_length(int year, int month) {
+  return (is_gregorian_leap_year(year) && month == 2)
+             ? 29
+             : non_leap_year_month_lengths[month - 1];
+}
+
+constexpr int mod_7_offset_1 (int const x) {
+  return x % 7 == 0 ? 7 : (x % 7);
+}
+
+// Result in [1, 7], 1 is Monday.
+constexpr int day_of_week_on_january_1st(int const year) {
+  // Gauss's formula, see
+  // https://en.wikipedia.org/wiki/Determination_of_the_day_of_the_week#Gauss.27s_algorithm.
+  return mod_7_offset_1(1 +
+                        5 * ((year - 1) % 4) +
+                        4 * ((year - 1) % 100) +
+                        6 * ((year - 1) % 400));
+}
+
+constexpr bool is_53_week_year(int const year) {
+  return day_of_week_on_january_1st(year) == 4 ||
+         (is_gregorian_leap_year(year) &&
+          day_of_week_on_january_1st(year) == 3);
+}
+
+constexpr int ordinal_of_w_01_1(int const year) {
+  return day_of_week_on_january_1st(year) <= 4
+             ? 2 - day_of_week_on_january_1st(year)
+             : (9 - day_of_week_on_january_1st(year));
+}
+
+constexpr int days_from_2000_01_01_at_start_of_2000_based_year(
+    int const years_from_2000) {
+  return years_from_2000 > 0
+             ? 1 + years_from_2000 * 365 +
+               (years_from_2000 - 1) / 4 -
+               (years_from_2000 - 1) / 100 +
+               (years_from_2000 - 1) / 400
+             : years_from_2000 * 365 +
+               years_from_2000 / 4 -
+               years_from_2000 / 100 +
+               years_from_2000 / 400;
+}
+
+// Returns the number formed by taking |size| increasingly significant digits,
```

Comment:
Somehow this feels backwards.  Maybe [begin, end) would work better.  For instance for YYYY-MM-DD I would expect to take (0, 4) (4, 6) and (6, 8).